frame/support/procedural/src/pallet/expand/event.rs

```diff
@@ -0,0 +1,143 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::pallet::Def;
+use syn::spanned::Spanned;
+
+/// * Add __Ignore variant on Event
+/// * Impl various trait on Event including metadata
+/// * if deposit_event is defined, implement deposit_event on module.
+pub fn expand_event(def: &mut Def) -> proc_macro2::TokenStream {
+	let event = if let Some(event) = &def.event {
+		event
+	} else {
+		return Default::default()
+	};
+
+	let event_where_clause = &event.where_clause;
+
+	// NOTE: actually event where clause must be a subset of config where clause because of
+	// `type Event: From<Event<Self>>`. But we merge either way for potential better error message
+	let completed_where_clause = super::merge_where_clauses(&[
+		&event.where_clause,
+		&def.config.where_clause,
+	]);
+
+	let event_ident = &event.event;
+	let frame_system = &def.frame_system;
+	let frame_support = &def.frame_support;
+	let event_use_gen = &event.event_use_gen();
+	let event_impl_gen= &event.event_impl_gen();
+	let metadata = event.metadata.iter()
+		.map(|(ident, args, docs)| {
+			let name = format!("{}", ident);
+			quote::quote!(
+				#frame_support::event::EventMetadata {
+					name: #frame_support::event::DecodeDifferent::Encode(#name),
+					arguments: #frame_support::event::DecodeDifferent::Encode(&[
+						#( #args, )*
+					]),
+					documentation: #frame_support::event::DecodeDifferent::Encode(&[
+						#( #docs, )*
+					]),
+				},
+			)
+		});
+
+	let event_item_span =
+		def.item.content.as_mut().expect("Checked by def parser").1[event.index].span();
+
+	let event_item = {
+		let item = &mut def.item.content.as_mut().expect("Checked by def parser").1[event.index];
+		if let syn::Item::Enum(item) = item {
+			item
+		} else {
+			unreachable!("Checked by event parser")
```

Comment:
Why can we not do the unpacking in the parser?

Reply:
I was thinking it was better to organize the code this way:
* the parser just retrieve informations from the module items.
* the expander just modifies the items and adds some generated code at the end.

This way I felt the user input will be best handled. For example if parser take their items then when they expand their stuff they may have to be careful that they are changing the order of user provided items.
But indeed it makes any modification of user input a bit annoying because we have to kind of parse them a second time.